test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/ServerSentEventsParserTests.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.Sockets.Internal.Formatters;
+using Xunit;
+using System.Threading.Tasks;
+using System.IO.Pipelines;
+using System;
+
+namespace Microsoft.AspNetCore.Sockets.Common.Tests.Internal.Formatters
+{
+    public class ServerSentEventsParserTests
+    {
+        [Theory]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\r\n", MessageType.Text, "Hello, World")]
+        [InlineData("data: E\r\ndata: Hello, World\r\n\r\n", MessageType.Error, "Hello, World")]
+        [InlineData("data: T\r\ndata: Hello\r\ndata: , World\r\n\r\n", MessageType.Text, "Hello, World")]
+        [InlineData("data: T\r\ndata: Major\r\ndata:  Key\r\ndata:  Alert\r\n\r\n", MessageType.Text, "Major Key Alert")]
+        [InlineData("data: T\r\n\r\n", MessageType.Text, "")]
+        public async Task ParseSSEMessageSuccessCases(string encodedMessage, MessageType messageType, string expectedMessage)
+        {
+            var stream = new MemoryStream();
+            var streamWriter = new StreamWriter(stream);
+            streamWriter.Write(encodedMessage);
+            streamWriter.Flush();
+            stream.Seek(0, SeekOrigin.Begin);
+
+            var pipelineReader = stream.AsPipelineReader();
+            var buffer = await pipelineReader.ReadToEndAsync();
+            var parser = new ServerSentEventsMessageParser();
+
+            var consumed = new ReadCursor();
+            var examined = new ReadCursor();
+
+            var parsePhase = parser.ParseMessage(buffer, out consumed, out examined, out Message message);
+            Assert.Equal(ServerSentEventsMessageParser.ParseResult.Completed, parsePhase);
+
+            var result = Encoding.UTF8.GetString(message.Payload);
+            Assert.Equal(expectedMessage, result);
+        }
+
+
+        [Theory]
+        [InlineData("", MessageType.Text, "Error parsing data from event stream")]
+        [InlineData("data: X\r\n\r\n", MessageType.Text, "Unknown message type: 'X'")]
+        [InlineData("data: X\r\n", MessageType.Text, "Unknown message type: 'X'")]
+        [InlineData("data:", MessageType.Text, "Expected a '\r\n' line ending")]
```

Comment:
This, as mentioned above in the implementation, isn't actually an error case. Any case where everything we've seen is **valid** but we just haven't seen enough to finish the frame is actually a case where we should return `Incomplete`, since we don't know if there's more data available.

Note the difference between the Incomplete and Invalid tests in the TextMessageParser: https://github.com/aspnet/SignalR/blob/dev/test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/TextMessageParserTests.cs#L84 - Feel free to ask me if you have questions about those tests or about what should be incomplete vs invalid in SSE.